### PR TITLE
first-run-wizard to open/create a workspace

### PR DESCRIPTION
**Pull request by ubruhin**
This pull request closes issue #10.

The new wizard to select a workspace is shown only the first time LibrePCB is started. For testing purposes you can remove the file ~/.config/LibrePCB/LibrePCB_git-first_run_wizard.conf. The next time LibrePCB is started, the wizard will be shown again.

The wizard could be easily extended by adding more pages (for example to choose some important workspace settings, like the measurement system).

@rnestler What do you think about the new wizard?

### CODE REVIEW

File: librepcb/firstrunwizard/firstrunwizard.cpp [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+#include "ui_firstrunwizard.h"
+#include "firstrunwizardpage_welcome.h"
+#include "firstrunwizardpage_workspacepath.h"
+
+/*****************************************************************************************
+ *  Constructors / Destructor
+ ****************************************************************************************/
+
+FirstRunWizard::FirstRunWizard(QWidget *parent) noexcept :
+    QWizard(parent), mUi(new Ui::FirstRunWizard)
```
Comment (rnestler) [inline librepcb/firstrunwizard/firstrunwizard.cpp]: This ``new/delete`` pairs should real`ly be replaced by ``std::unique_ptr<T>``. If this class owns the ``Ui::FirstRunWizard`` a unique pointer is perfectly suited.

File: librepcb/firstrunwizard/firstrunwizardpage_welcome.cpp [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+#include "firstrunwizardpage_welcome.h"
+#include "ui_firstrunwizardpage_welcome.h"
+
+/*****************************************************************************************
+ *  Constructors / Destructor
+ ****************************************************************************************/
+
+FirstRunWizardPage_Welcome::FirstRunWizardPage_Welcome(QWidget *parent) noexcept :
+    QWizardPage(parent), mUi(new Ui::FirstRunWizardPage_Welcome)
```
Comment (rnestler) [inline librepcb/firstrunwizard/firstrunwizardpage_welcome.cpp]: Same here, use ```std::unique_ptr<T>``

File: librepcb/firstrunwizard/firstrunwizardpage_workspacepath.cpp [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+#include "ui_firstrunwizardpage_workspacepath.h"
+#include <librepcbcommon/fileio/filepath.h>
+#include <librepcbworkspace/workspace.h>
+
+/*****************************************************************************************
+ *  Constructors / Destructor
+ ****************************************************************************************/
+
+FirstRunWizardPage_WorkspacePath::FirstRunWizardPage_WorkspacePath(QWidget *parent) noexcept :
+    QWizardPage(parent), mUi(new Ui::FirstRunWizardPage_WorkspacePath)
```
Comment (rnestler) [inline librepcb/firstrunwizard/firstrunwizardpage_workspacepath.cpp]: Another candidate for ``std::unique_ptr<T>``

**Comment by rnestler** [comment]
You should really start using smart pointers. Every time you write a destructor you should ask yourself if you're doing something wrong.

**Comment by 0xB767B** [comment]
I agree with raffi. remo too.

**Comment by ubruhin** [comment]
The UI is the only object which is deleted in the destructor, and this code was automatically generated by qtcreator on creating the new class (but slightly modified by me afterwards). You can replace the raw pointer by a smart pointer but this costs time which could be used for more important tasks...

But you're right, there are other classes which use lot of raw pointers where smart pointers would make more sense... I will try to use more smart pointers in future, but the problem is that i really like raw pointers xD

**Comment by ubruhin** [comment]
better? :smile:

**Comment by rnestler** [comment]
@ubruhin I have yet to look at the new wizard :wink: and it's much more fun to nag about code.

**Comment by ubruhin** [comment]
@rnestler Completely removing the destructor is not possible since the default destructor would then be inline, and the type of the QScopedPointer template is only forward declared. So I have explicitly defined the empty destructor.

And what's wrong with QScopedPointer? :smile: Are there any advantages by using the STL pointer types?

**Comment by rnestler** [comment]
A forward declared template really?
Regarding advantages of using the STL types: Have you looked at the link I provided? ``QScopedPointer`` does not provide move semantics for example.

**Comment by ubruhin** [comment]
Not the template itself is forward declared, but the type of the object we are pointing to. From firstrunwizard.h:
```
QScopedPointer<Ui::FirstRunWizard> mUi;
```
The class `Ui::FirstRunWizard` is only a forward declaration, so an inline destructor does not work as the compiler does not know the destructor of `Ui::FirstRunWizard`. In firstrunwizard.cpp we include ui_firstrunwizard.h, so the declaration of the class `Ui::FirstRunWizard` is available.

This is always a problem when using smart pointers for member variables of forward declared types...

Move semantics are not really required as the following code should do the same:
```
QScopedPointer<Ui::FirstRunWizard> a(new Ui::FirstRunWizard());
QScopedPointer<Ui::FirstRunWizard> b(a.take());
```
But which pointers (Qt or STL) we use should not really matter, as long as we use the same for the whole project. And I have already used Qt's smart pointers before this pull requesr ;-)

**Comment by rnestler** [comment]
Currently it doesn't build for me:
```
make[1]: Leaving directory '/home/roughl/proggen/build-librepcb-Desktop-Debug/tools'
../generated/unix/liblibrepcbproject.a(netsignal.o): In function `project::NetSignal::NetSignal(project::Circuit const&, XmlDomElement const&)':
/home/roughl/proggen/build-librepcb-Desktop-Debug/libs/librepcbproject/../../../LibrePCB/libs/librepcbproject/circuit/netsignal.cpp:48: undefined reference to `vtable for project::NetSignal'
/home/roughl/proggen/build-librepcb-Desktop-Debug/libs/librepcbproject/../../../LibrePCB/libs/librepcbproject/circuit/netsignal.cpp:48: undefined reference to `vtable for project::NetSignal'
/home/roughl/proggen/build-librepcb-Desktop-Debug/libs/librepcbproject/../../../LibrePCB/libs/librepcbproject/circuit/netsignal.cpp:48: undefined reference to `vtable for project::NetSignal'
../generated/unix/liblibrepcbproject.a(netsignal.o): In function `project::NetSignal::NetSignal(project::Circuit const&, project::NetClass&, QString const&, bool)':
/home/roughl/proggen/build-librepcb-Desktop-Debug/libs/librepcbproject/../../../LibrePCB/libs/librepcbproject/circuit/netsignal.cpp:68: undefined reference to `vtable for project::NetSignal'
/home/roughl/proggen/build-librepcb-Desktop-Debug/libs/librepcbproject/../../../LibrePCB/libs/librepcbproject/circuit/netsignal.cpp:68: undefined reference to `vtable for project::NetSignal'
../generated/unix/liblibrepcbproject.a(netsignal.o):/home/roughl/proggen/build-librepcb-Desktop-Debug/libs/librepcbproject/../../../LibrePCB/libs/librepcbproject/circuit/netsignal.cpp:68: more undefined references to `vtable for project::NetSignal' follow
../generated/unix/liblibrepcbproject.a(netsignal.o): In function `project::NetSignal::setName(QString const&, bool)':
/home/roughl/proggen/build-librepcb-Desktop-Debug/libs/librepcbproject/../../../LibrePCB/libs/librepcbproject/circuit/netsignal.cpp:101: undefined reference to `project::NetSignal::nameChanged(QString const&)'
../generated/unix/liblibrepcbproject.a(netsignal.o): In function `project::NetSignal::tr(char const*, char const*, int)':
/home/roughl/proggen/build-librepcb-Desktop-Debug/libs/librepcbproject/../../../LibrePCB/libs/librepcbproject/circuit/netsignal.h:56: undefined reference to `project::NetSignal::staticMetaObject'
../generated/unix/liblibrepcbproject.a(gencompsignalinstance.o): In function `project::GenCompSignalInstance::init()':
/home/roughl/proggen/build-librepcb-Desktop-Debug/libs/librepcbproject/../../../LibrePCB/libs/librepcbproject/circuit/gencompsignalinstance.cpp:90: undefined reference to `project::NetSignal::nameChanged(QString const&)'
../generated/unix/liblibrepcbproject.a(gencompsignalinstance.o): In function `project::GenCompSignalInstance::setNetSignal(project::NetSignal*)':
/home/roughl/proggen/build-librepcb-Desktop-Debug/libs/librepcbproject/../../../LibrePCB/libs/librepcbproject/circuit/gencompsignalinstance.cpp:128: undefined reference to `project::NetSignal::nameChanged(QString const&)'
/home/roughl/proggen/build-librepcb-Desktop-Debug/libs/librepcbproject/../../../LibrePCB/libs/librepcbproject/circuit/gencompsignalinstance.cpp:137: undefined reference to `project::NetSignal::nameChanged(QString const&)'
../generated/unix/liblibrepcbproject.a(gencompsignalinstance.o): In function `QMetaObject::Connection QObject::connect<void (project::NetSignal::*)(QString const&), void (project::GenCompSignalInstance::*)(QString const&)>(QtPrivate::FunctionPointer<void (project::NetSignal::*)(QString const&)>::Object const*, void (project::NetSignal::*)(QString const&), QtPrivate::FunctionPointer<void (project::GenCompSignalInstance::*)(QString const&)>::Object const*, void (project::GenCompSignalInstance::*)(QString const&), Qt::ConnectionType)':
/usr/include/qt/QtCore/qobject.h:239: undefined reference to `project::NetSignal::staticMetaObject'
../generated/unix/liblibrepcbproject.a(gencompsignalinstance.o): In function `bool QObject::disconnect<void (project::NetSignal::*)(QString const&), void (project::GenCompSignalInstance::*)(QString const&)>(QtPrivate::FunctionPointer<void (project::NetSignal::*)(QString const&)>::Object const*, void (project::NetSignal::*)(QString const&), QtPrivate::FunctionPointer<void (project::GenCompSignalInstance::*)(QString const&)>::Object const*, void (project::GenCompSignalInstance::*)(QString const&))':
/usr/include/qt/QtCore/qobject.h:382: undefined reference to `project::NetSignal::staticMetaObject'
../generated/unix/liblibrepcbproject.a(si_netlabel.o): In function `project::SI_NetLabel::init()':
/home/roughl/proggen/build-librepcb-Desktop-Debug/libs/librepcbproject/../../../LibrePCB/libs/librepcbproject/schematics/items/si_netlabel.cpp:68: undefined reference to `project::NetSignal::nameChanged(QString const&)'
../generated/unix/liblibrepcbproject.a(si_netlabel.o): In function `project::SI_NetLabel::setNetSignal(project::NetSignal&)':
/home/roughl/proggen/build-librepcb-Desktop-Debug/libs/librepcbproject/../../../LibrePCB/libs/librepcbproject/schematics/items/si_netlabel.cpp:100: undefined reference to `project::NetSignal::nameChanged(QString const&)'
/home/roughl/proggen/build-librepcb-Desktop-Debug/libs/librepcbproject/../../../LibrePCB/libs/librepcbproject/schematics/items/si_netlabel.cpp:104: undefined reference to `project::NetSignal::nameChanged(QString const&)'
../generated/unix/liblibrepcbproject.a(si_netlabel.o): In function `QMetaObject::Connection QObject::connect<void (project::NetSignal::*)(QString const&), void (project::SI_NetLabel::*)(QString const&)>(QtPrivate::FunctionPointer<void (project::NetSignal::*)(QString const&)>::Object const*, void (project::NetSignal::*)(QString const&), QtPrivate::FunctionPointer<void (project::SI_NetLabel::*)(QString const&)>::Object const*, void (project::SI_NetLabel::*)(QString const&), Qt::ConnectionType)':
/usr/include/qt/QtCore/qobject.h:239: undefined reference to `project::NetSignal::staticMetaObject'
../generated/unix/liblibrepcbproject.a(si_netlabel.o): In function `bool QObject::disconnect<void (project::NetSignal::*)(QString const&), void (project::SI_NetLabel::*)(QString const&)>(QtPrivate::FunctionPointer<void (project::NetSignal::*)(QString const&)>::Object const*, void (project::NetSignal::*)(QString const&), QtPrivate::FunctionPointer<void (project::SI_NetLabel::*)(QString const&)>::Object const*, void (project::SI_NetLabel::*)(QString const&))':
/usr/include/qt/QtCore/qobject.h:382: undefined reference to `project::NetSignal::staticMetaObject'
collect2: error: ld returned 1 exit status
make[1]: *** [../generated/unix/librepcb] Error 1
make: *** [sub-librepcb-make_first] Error 2
Makefile:250: recipe for target '../generated/unix/librepcb' failed
make[1]: Leaving directory '/home/roughl/proggen/build-librepcb-Desktop-Debug/librepcb'
Makefile:96: recipe for target 'sub-librepcb-make_first' failed
```
Maybe there is some dependency missing?

**Comment by ubruhin** [comment]
@rnestler You need to run qmake again (or simply remove the whole output directory), see this Stack Overflow answer: http://stackoverflow.com/questions/2555816/qt-linker-error-undefined-reference-to-vtable

The problem is caused by commit 9e55d48d123e7e3a100c381534556cafb7e1315b where I have added QObject as a parent class of project::NetClass.

File: librepcb/librepcb.pro [hunk 4]
```diff
@@ -15,7 +15,7 @@ include(../common.pri)
 
 QT += core widgets opengl webkitwidgets xml printsupport sql
 
-exists(../.git):DEFINES += GIT_BRANCH=\\\"master\\\"
+exists(../.git):DEFINES += GIT_BRANCH=\\\"first_run_wizard\\\"
```
Comment (rnestler) [inline librepcb/librepcb.pro]: Out of curiosity: What's that for? Do you have to change that manually?

Comment (ubruhin) [inline librepcb/librepcb.pro]: haha :) At the beginning, I have used the line:
```
exists(.git):DEFINES += GIT_BRANCH=\\\"$$system(git rev-parse --abbrev-ref HEAD)\\\"
```
But on "some operating systems" this did not work, so I simply have replaced the command with a hardcoded string later...

**Pull request closed without merging**